File: Projects/eigen.vb
Option Base 1


Dim k As Integer, l As Integer, n As Integer
Dim rho As Double, accuracy As Double, MaxIter As Integer
Dim ValuesDifferent As Boolean

Public matrix As Variant
Public Eigenvectors As Variant
Public Eigenvalues As Variant
Public tempmax As Double
Dim j() As Double


Private Sub Class_Initialize()
    accuracy = 0.00001
    MaxIter = 500
End Sub

Public Sub Init(E As Variant, Optional accuracy_ As Variant, Optional MaxIter_ As Variant, Optional bolSort As Boolean)
    matrix = E
    n = UBound(E, 1)
    ReDim j(n, n)
    ReDim Eigenvalues(n)
    If Not IsMissing(accuracy_) Then accuracy = accuracy_
    If Not IsMissing(MaxIter_) Then MaxIter = MaxIter_
    ValuesDifferent = True
    Call calculate
    If bolSort = True Then Call Sort
    Eigenvalues = Application.Transpose(Eigenvalues) 'eigenvalues are a row vector before transposing
End Sub

Private Sub maxind()
'Get the largest below diagonal component

    Dim R As Integer, c As Integer
    tempmax = -1
    
    For R = 2 To n
        For c = 1 To R - 1
            If Abs(matrix(R, c)) > tempmax Then
                tempmax = Abs(matrix(R, c))
                k = R: l = c
            End If
        Next c
    Next R

End Sub

Private Sub CalcRho()
'Calculate the value of rho
    
    If matrix(k, k) <> matrix(l, l) Then
        rho = Atn(2 * matrix(k, l) / (matrix(k, k) - matrix(l, l))) / 2
    Else
        rho = Application.Pi * Sgn(matrix(k, l)) / 4
    End If
    
End Sub

Private Sub JacobiMatrix(i As Integer)
'Determine the matris for the jacobi rotation

    j = identity(n)
    j(k, k) = Cos(rho)
    j(k, l) = -Sin(rho)
    j(l, k) = Sin(rho)
    j(l, l) = Cos(rho)
    If i > 1 Then
        Eigenvectors = Application.MMult(Eigenvectors, j)
    Else
        Eigenvectors = j
    End If
End Sub

Private Sub calculate()
'calculate the matrix with the diagonals as the eigenvalues

    Dim i As Integer
    i = 0
    Do While i <= MaxIter And ValuesDifferent
        i = i + 1
        Call maxind
        Call CalcRho
        Call JacobiMatrix(i)
        matrix = Application.MMult(Application.MMult(Application.Transpose(j), matrix), j)
        Call CheckAccuracy
    Loop
End Sub

Private Function identity(dimension As Integer) As Variant
'Create an identity matrix
    Dim A() As Double
    ReDim A(dimension, dimension)
    Dim index As Integer
    For index = 1 To dimension
        A(index, index) = 1
    Next index
    identity = A
End Function

Private Sub ExtractEigenvalues()
    Dim i As Integer
    For i = 1 To n
        Eigenvalues(i) = matrix(i, i)
    Next i
End Sub

Private Sub CheckAccuracy()
'See if eigenvalues are significant
    Dim i As Integer
    ValuesDifferent = False
    If Abs(tempmax) > accuracy Then ValuesDifferent = True
    ExtractEigenvalues
End Sub

Public Function getAccuracy()
    getAccuracy = accuracy
End Function
Public Sub setAccuracy(accuracy_ As Double)
    accuracy = accuracy_
End Sub
Public Function getMaxIter()
    getMaxIter = MaxIter
End Function
Public Sub setMaxIter(MaxIter_ As Integer)
    MaxIter = MaxIter_
End Sub

Private Sub Sort()
    Dim i As Integer, passNum As Integer
    For passNum = 1 To n - 1
        For i = 1 To n - passNum
            If Eigenvalues(i) < Eigenvalues(i + 1) Then
                Call VectorEntrySwap(Eigenvalues, i, i + 1)
                Call ColSwap(Eigenvectors, i, i + 1)
            End If
        Next i
    Next passNum
End Sub

Private Sub VectorEntrySwap(vector As Variant, entry1 As Integer, entry2 As Integer)
'Swap entries within a vector
    Dim temp As Variant
    temp = vector(entry1)
    vector(entry1) = vector(entry2)
    vector(entry2) = temp
End Sub

Private Sub ColSwap(matrix As Variant, col1 As Integer, col2 As Integer)
'Swap the columns within a vector
    Dim temp As Variant
    Dim i As Integer
    For i = LBound(matrix, 1) To UBound(matrix, 1)
        temp = matrix(i, col1)
        matrix(i, col1) = matrix(i, col2)
        matrix(i, col2) = temp
    Next i
End Sub

File: Projects/eigen.php
<?php
/*****************************************************************************************************************
** File: eigen.php
* Parent: calc_eigen.php
* This file creates a class called Eigen which contains functions particularly:  linear algebra calculations.
*
*****************************************************************************************************************/
class Eigen
{
	/*Constructor*/
	function __construct()
	{
		$this->accuracy = 0.00001;
		$this->Maxlter = 500;
	}
	
	/*Destructor*/
	function __destruct()
	{
		unset($this);
	}
	
	public function Init(array $E)
	{
		$this->matrix = $E;
		$this->n = count($E);

		$this->ValuesDifferent = True; //Used for accuracy checking
		$this->calculate();
		// Transpose to column vector because EigenValues are a row vector before transposing.
		// and sort Eigenvalues Descending in value
		$j=1;
		
		if($this->EigenValues[0] < $this->EigenValues[1]) 
		{
			for($i=0; $i < $this->n; $i++)
			{
				$this->EigenValuesCol[$i] = $this->EigenValues[$j];
				$j--;
			}
		}
		else
		{
			for($i=0; $i < $this->n; $i++)
			{
				$this->EigenValuesCol[$i] = $this->EigenValues[$i];
			}
		}
		
		// Transpose EigenVectors
		for($j=0; $j < $this->n; $j++)
		{
			for($k=0; $k < $this->n; $k++)
			{
				$this->Transposed_Eigen_Vector[$k][$j] = $this->EigenVectors[$j][$k];
			}
		}
	}
	
	/* Calculate the matrix with the diagonals as the EigenValues */
	private function calculate()
	{
		$i = 0;
		do 
		{
			$i = $i +1;
			$this->maxind();
			$this->CalcRho();
			$this->JacobiMatrix($i);
			// Transpose Identity Matrix (to be used below)
			for($j=0; $j < $this->n; $j++)
			{
				for($k=0; $k < $this->n; $k++)
				{
					$this->Transposed_Identity[$j][$k] = $this->j[$k][$j];
				}
			}
			// Original Matrix =  Identity Matrix Transposed * Original Matrix * Identity Matrix
			$this->matrix = $this->matrix_multiplication( $this->matrix_multiplication($this->Transposed_Identity, $this->matrix) , $this->j );

			$this->CheckAccuracy();
		}while ($i <= $this->Maxlter && $this->ValuesDifferent);
	}
	
	/* Get the largest below diagonal component */
	private function maxind()
	{
		$this->tempmax = -1;
		for ($R=1; $R<=$this->n-1; $R++)
		{
			for ($c=0; $c<=$R-1; $c++)
			{
				if ( abs($this->matrix[$R][$c]) > $this->tempmax )
				{
					$this->tempmax = abs($this->matrix[$R][$c]);
					$this->k = $R;
					$this->l = $c;
				}
			}
		}
	}
	/* Calculate the value of rho */
	private function CalcRho()
	{
		if ( $this->matrix[$this->k][$this->k] != $this->matrix[$this->l][$this->l] )
		{
			//ArcTangent
			$this->rho = atan( 2 * $this->matrix[$this->k][$this->l] / ($this->matrix[$this->k][$this->k] - $this->matrix[$this->l][$this->l]) ) / 2;
		}
		else
		{
			$this->rho = pi() * $this->sign($this->matrix[$this->k][$this->l]) / 4;
		}
	}
	
	private function sign( $number ) 
	{ 
		return ( $number > 0 ) ? 1 : ( ( $number < 0 ) ? -1 : 0 ); 
	}	 
	
	private function JacobiMatrix($i)
	{
		//Set $j to identity Matrix
		$this->j = $this->identity($this->n);
		//Assign new values to $j
		$this->j[$this->k][$this->k] = Cos($this->rho);
		$this->j[$this->k][$this->l] = -Sin($this->rho);
		$this->j[$this->l][$this->k] = Sin($this->rho);
		$this->j[$this->l][$this->l] = Cos($this->rho);
		//Calculate EigenVectors
		if ($i >1 )
			$this->EigenVectors = $this->matrix_multiplication($this->EigenVectors, $this->j);
		else
			$this->EigenVectors = $this->j;
	}
	
	public function matrix_multiplication($matrix_A, $matrix_B)
	{
		for ($i=0; $i < $this->n; $i++)
		{
			for ($j=0; $j < $this->n; $j++)
			{
				$matrix_C[$i][$j] = 0; //Loops through rows, then columns to initialize all elements to zero
				for($k=0; $k < $this->n; $k++)
				{
				 $matrix_C[$i][$j] += $matrix_A[$i][$k] * $matrix_B[$k][$j];
				}
			}
		}
		return $matrix_C;
	}
	
	/* Creates an Identity Matrix by filling diagonal with 1's, rest of matrix should be zeroes */
	private function identity($dimension)
	{
		for ($i = 0; $i < $dimension ; $i++)
		{
			for ($j=0; $j < $dimension; $j++)
			{
				$Array[$i][$j] = 0;
				if ($i == $j)
					$Array[$i][$j] = 1;
			}
		}
		return $Array;
	}
	
	/* See if EigenValues are significant */
	private function CheckAccuracy()
	{
		$this->ValuesDifferent = False;

		If (abs($this->tempmax) > $this->accuracy)
		{
			$this->ValuesDifferent = True;
		}
		$this->ExtractEigenValues();
	}
	
	/* Grab diagonal from matrix */
	private function ExtractEigenValues()
	{
		for ($i=0;$i < $this->n; $i++)
		{
			$this->EigenValues[$i] = $this->matrix[$i][$i];
		}
	}
}

?>